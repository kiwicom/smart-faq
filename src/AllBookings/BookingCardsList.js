// @flow

import idx from 'idx';
import * as React from 'react';
import { graphql, createFragmentContainer } from 'react-relay';
import { Heading } from '@kiwicom/orbit-components';

import OneWayBooking from './BookingTypes/OneWayBooking';
import ReturnBooking from './BookingTypes/ReturnBooking';
import MulticityBooking from './BookingTypes/MulticityBooking';
import type { BookingCardsList_booking } from './__generated__/BookingCardsList_booking.graphql';
import { BookingState, type BookingStateType } from '../context/BookingState';

type Props = {
  booking: BookingCardsList_booking,
  title: string,
};

const BookingCardsList = (props: Props) => {
  const { booking, title } = props;
  const edges = idx(booking, _ => _.edges) || [];
  const bookings = edges.map(edge => idx(edge, _ => _.node));

  return (
    <React.Fragment>
      {bookings.length > 0 && (
        <div className="subtitle">
          <Heading weight="medium" size="small">
            {title}
          </Heading>
        </div>
      )}
      {bookings.map(booking => {
        const id = idx(booking, _ => _.id);
        const type = idx(booking, _ => _.__typename);
        const variants: Object = {
          BookingOneWay: <OneWayBooking booking={booking} />,
          BookingReturn: <ReturnBooking booking={booking} />,
          BookingMulticity: <MulticityBooking booking={booking} />,
        };

        let bookingComponent = null;
        if (variants[type]) {
          bookingComponent = variants[type];
        }

        if (bookingComponent) {
          if (!id) {
            return bookingComponent;
          }

          return (
            <BookingState.Consumer key={id}>
              {({ onSelectBooking }: BookingStateType) => (
                <div
                  onClick={() => onSelectBooking(id)}
                  className="bookingCard"
                  role="button"
                  onKeyUp={() => onSelectBooking(id)}
                  tabIndex={0}
                >
                  {bookingComponent}
                </div>
              )}
            </BookingState.Consumer>
          );
        }

        return null;
      })}
      <style jsx>
        {`
          .bookingCard {
            cursor: pointer;
            outline: none;
          }
          div.subtitle {
            margin-top: 14px;
            margin-bottom: 12px;
          }
        `}
      </style>
    </React.Fragment>
  );
};

export const RawBookingCardsList = BookingCardsList;

export default createFragmentContainer(
  BookingCardsList,
  graphql`
    fragment BookingCardsList_booking on BookingInterfaceConnection {
      edges {
        node {
          id
          __typename
          ... on BookingOneWay {
            ...OneWayBooking_booking
          }
          ... on BookingReturn {
            ...ReturnBooking_booking
          }
          ... on BookingMulticity {
            ...MulticityBooking_booking
          }
        }
      }
    }
  `,
);
