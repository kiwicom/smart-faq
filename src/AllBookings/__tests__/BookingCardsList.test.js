// @flow
import * as React from 'react';
import { shallow } from 'enzyme';

import { RawBookingCardsList } from '../BookingCardsList';

const mockRefType: any = null;
describe('RawBookingCardsList', () => {
  it('no bookings in list', () => {
    const booking = {
      edges: [],
      $refType: mockRefType,
    };
    const result = shallow(
      <RawBookingCardsList title="Example" booking={booking} />,
    );
    expect(result).toMatchSnapshot();
  });

  it('bookings in list', () => {
    const booking = {
      edges: [{
        node: null,
      }],
      $refType: mockRefType,
    };
    const result = shallow(
      <RawBookingCardsList title="Example" booking={booking} />,
    );
    expect(result).toMatchSnapshot();
  });
});
