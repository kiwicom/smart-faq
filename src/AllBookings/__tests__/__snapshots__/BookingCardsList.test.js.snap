// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RawBookingCardsList bookings in list 1`] = `
<React.Fragment>
  <div
    className="jsx-1893935473 subtitle"
  >
    <Heading
      element="h1"
      size="small"
      weight="medium"
    >
      Example
    </Heading>
  </div>
  <JSXStyle
    css=".bookingCard.jsx-1893935473{cursor:pointer;outline:none;}div.subtitle.jsx-1893935473{margin-top:14px;margin-bottom:12px;}"
    styleId="1893935473"
  />
</React.Fragment>
`;

exports[`RawBookingCardsList no bookings in list 1`] = `
<React.Fragment>
  <JSXStyle
    css=".bookingCard.jsx-1893935473{cursor:pointer;outline:none;}div.subtitle.jsx-1893935473{margin-top:14px;margin-bottom:12px;}"
    styleId="1893935473"
  />
</React.Fragment>
`;
